### DISCUSSION TITLE
Router rules and channel muting

### DISCUSSION THREAD

**Discussion by Stratg** [GeekFunkLabs/fluidpatcher#106] [Q&A]
Hi,

I have an axiom 61 which has 9 faders. I would like to be able to configure 9 midi channels each with an instrument, and use the faders to mix these instruments (maximum 4 or 5 simultaneously).

The 9 faders currently control the expression (CC11). However, even with the volume of some faders at 0, fluidsynth plays the samples.
The power of the raspberry pi 3 does not allow me to read these 9 instruments simultaneously, I am blocked either by the CPU or by the polyphony.
On the other hand, a key played with a velocity of 0 does not trigger a process.

I've thought about a few ideas and so I have a few questions regarding the routing rules:

1 - Is it possible to use a fader to clip the velocity of the notes played? for example, at position 50, notes played with a velocity between 0 and 50 would sound normally, but above 50, notes would be limited to 50 (a bit like an automobile speed limiter). At zero, the channel would therefore be inert.

2 - Is it possible to mute a particular midi channel by pressing a CC toggle button (values ​​0 or 127)?

3 - Is it possible to change the routing of a midi channel with a rule? For example send notes from a channel that I want to mute to an unused channel (like channel 16)

Or do you have another idea ?

Thank you for your help !

**Comment by Stratg** [comment]
Or perhaps editing python script... But i'can't find the Fluidpatcher API reference.    "fluidpatcher/docs/api_reference.md" is empty...

Thanks !!

**Reply by albedozero** [reply]
Here's an example of how this _should_ work - if it doesn't or you already tried let me know. I just did it for 4 different channels to make things less cluttered. This assumes your keyboard sends notes on channel 1, and that you have 4 faders that send CC's 12-15 (good general use CCs) on channel 1.

```yaml
patches:
  PresetMixer:
    2: defaultGM.sf2:000:000
    3: defaultGM.sf2:000:001
    4: defaultGM.sf2:000:002
    5: defaultGM.sf2:000:003
router_rules:
  - {type: cc, chan: 1=2, par1: 12=7}
  - {type: cc, chan: 1=3, par1: 13=7}
  - {type: cc, chan: 1=4, par1: 14=7}
  - {type: cc, chan: 1=5, par1: 15=7}
```

I routed the sliders to CC7 (volume) instead of CC11 - either works. You could also just program the faders on your Axiom to send CC 7 or 11 on channels 2-5 instead of using the router rules. Notes sent to channels with no expression/volume _should_ generate no voices and thus not tax the CPU.

On your router rule questions:

1. You can clip note velocities with the `par2` router rule parameter like so:

```yaml
- {type: note, chan: 1=2, par2: 0-50}
- {type: note, chan: 1=2, par2: 51-127=50}
```

2. Sure, you can use volume (CC7) or expression (CC11) to mute - the output of the channel is multiplied by the product of those two values. I usually use CC11 to mute channels, and CC7 to adjust levels with a fader - that way when I unmute the volume is back where I set it rather than jumping to max. For example, if you have a toggle button that sends CC 16 on channel 1, you could add the rule `{type: cc, chan: 1=2, par1: 16=11}` to the example above.

3. Router rules cannot modify other router rules. I've thought about implementing this, but it would be a mess and I've haven't really found a case where I can't find another way to do something I want.

I've been updating docs and finally got the official build [posted](https://geekfunklabs.github.io/fluidpatcher) (including pretty API reference stuff).

**Reply by Stratg** [reply]
Hi,
Thanks for your time and the examples above.
Unfortunately, expression and volume at zero still generate voices. This is a "test" patch which generate high cpu stress:

```
  TestCpu:
    2: Stratagem.sf2:000:000 # 
    3: Stratagem.sf2:000:001 # 
    4: Stratagem.sf2:000:002 # 
    5: Stratagem.sf2:000:003 # 
    6: Stratagem.sf2:000:004 # 
    7: Stratagem.sf2:000:005 # 
    8: _Stratagem_Orchestre.sf2:000:000 # 
    9: _Stratagem_Orchestre.sf2:000:001 #   
    10: _Stratagem_Orchestre.sf2:000:002 # 
    11: _Stratagem_Orchestre.sf2:000:003 # 
    
    router_rules:
    - {type: note, chan: 1=2-11}
    - {type: cc, chan: 1=2-11, par1: 74=7}  #Fader 1
    - {type: cc, chan: 1=2-11, par1: 71=11} #Fader 2
```
Even with expression and volume at zero, we can see there is a CPU load when pressing many keys:
![cpu_high](https://github.com/GeekFunkLabs/fluidpatcher/assets/108470745/3c43edc0-1113-422b-9c00-ca555906e87b)

For comparison, when no note are played, this is the CPU usage:
![cpu_down](https://github.com/GeekFunkLabs/fluidpatcher/assets/108470745/642715b9-f47b-4f93-8144-9635ae9ea49e)

I will think differently about the construction of my patches and the use of my keyboard.

Thank you for API reference.

**Reply by albedozero** [reply]
Hmm. I've done some more tests of my own, even using FluidSynth directly, and I'm seeing the same thing - dropping MIDI expression or volume to zero doesn't affect the CPU load when many notes are being sustained. It must be that the attenuation is applied after the audio is generated. I'm not sure if this is the way FluidSynth _has_ to work, or if there is a way to optimize it. I'll have to think about that, and I'll also try to think of a better way to mute voices.

At the same time, there could be some other reasons for your heavy CPU load. Are you using an older Pi 3 perhaps (as in not the 3B+), or is the soundfont you're using extremely large or high quality?

**Comment by Stratg** [comment]
Thanks! I use a PI3B (not +) which has not a really powerfull CPU. Indeed, the fonts that I use are very large (between 40 and 80 MB) which also explains a high processor load when I apply many patches simultaneously. Moreover, theses fonts does not allow the use  laspa effect such as overdrive or a rotay speaker. This collapses the CPU.  I could lighten the fonts when I want to use a large number of channels.

To mute a channel and not trigger voices, one option would be to limit the note velocity to zero, but this does not seem to be possible from a CC command as you explained to me above.
